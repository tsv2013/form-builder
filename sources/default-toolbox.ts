import { createPropertyEditor, makeTitle } from "./utils";

export const defaultToolboxItems = [
    {
        title: "row",
        hint: "Drag to add row",
        json: {
            partclass: "layoutRow",
            cssClasses: "row",
            parts: [

            ]
        }
    },
    {
        title: "column",
        hint: "Drag to add column",
        json: {
            partclass: "layoutColumn",
            cssClasses: "column",
            parts: [

            ]
        }
    },
    {
        title: "panel",
        hint: "Drag to add panel",
        json: {
            partclass: "panel",
            cssClasses: "panel",
            parts: [

            ]
        }
    },
    {
        title: "label",
        hint: "Drag to add label",
        json: {
            partclass: "label",
            cssClasses: "test-label",
            data: "Label text"
        }
    },
    {
        title: "input",
        hint: "Drag to add input element",
        json: {
            partclass: "input",
            cssClasses: "test-input",
            data: "valName"
        }
    },
    {
        title: "composite",
        hint: "Drag to add composite element",
        json: {
            partclass: "layoutItem",
            cssClasses: "item test-item test-group",
            parts: [
                {
                    partclass: "label",
                    cssClasses: "test-label",
                    data: "Name"
                },
                {
                    partclass: "input",
                    cssClasses: "test-input",
                    data: "name"
                }
            ]
        }
    }
];

export function createToolboxItemsFor(model: any) {
    return (Object.keys(model) || [])
        .filter(propertyName => typeof model[propertyName] !== "object")
        .map(propertyName => {
            return {
                title: makeTitle(propertyName),
                hint: "Drag to add " + makeTitle(propertyName),
                json: createPropertyEditor(propertyName)
            };
        });
}

export function createDefaultToolboxItems(model: any) {
    return [].concat(defaultToolboxItems).concat(createToolboxItemsFor(model));
}