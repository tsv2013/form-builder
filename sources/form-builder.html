<div>
    <div>
        <div class="bf-toolbox-item" draggable="true" data-bind="event: { dragstart: dragstart }">

        </div>
    </div>
    <div>
        <layout-item params="item: root"></layout-item>
    </div>
</div>