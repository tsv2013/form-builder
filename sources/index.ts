export * from "./form-builder";
export * from "./uiml-parts";
export * from "./form-element";
export * from "./layout-item";
export * from "./placeholder-item";
export * from "./item-menu";
export * from "./utils";
export * from "./default-toolbox";
export * from "./form";
