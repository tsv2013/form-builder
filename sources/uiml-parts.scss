.row {
    min-height: 100px;
}
.column {
    display: inline-block;
}