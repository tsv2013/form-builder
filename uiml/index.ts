import { IUML, IUMLInterface, IPart, htmlUiRendererPeers } from "./uiml";
import { UIMLWrapper } from "./uiml-wrapper";

export * from "./uiml";
export * from "./renderer";
export * from "./panel";
export * from "./layout";
export * from "./object";
export * from "./collection";

declare var config: IUML;

const uimlWrapper = new UIMLWrapper(typeof config !== "undefined" && config || { uimlinterface: <IUMLInterface>{}, peers: htmlUiRendererPeers });

export function render(part: IPart, container?: HTMLElement) {
    const renderer = uimlWrapper.createRenderer(part);
    return renderer.render(container);
}

export function refresh() {
    uimlWrapper.refresh();
}