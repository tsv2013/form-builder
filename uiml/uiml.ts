export interface IPart {
    id?: string;
    partclass: string;
    parts?: IPart[];
    cssClasses?: string;
    [others: string]: any;
}

export interface IStyle {
    partname?: string;
    partclass?: string;
    name: string;
    value: string;
}

export interface IUMLInterface {
    parts: IPart[];
    styles?: IStyle[];
}

export interface IProperty {
    id: string;
    mapsto: string;
}

export interface IClass {
    id: string;
    mapsto: string;
    dproperties?: IProperty[];
}

export interface IPeer {
    peername: string;
    id: string;
    dclasses: IClass[];
    wloaders: { [widget: string]: () => Promise<any> };
}

export interface IUML {
    uimlinterface: IUMLInterface;
    peers: IPeer[];
}

export let htmlUiRendererPeers = [
    <IPeer>{
        peername: "presentation",
        id: "HTMLUIRenderer",
        dclasses: [
            { id: "container", mapsto: "ContainerRenderer" },
            { id: "input", mapsto: "InputRenderer" },
            { id: "label", mapsto: "LabelRenderer" }
        ],
        wloaders: <{ [widget: string]: () => Promise<any> }>{}
    }
];
