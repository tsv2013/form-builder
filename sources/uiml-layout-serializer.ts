import { FormElement, IFormElement } from "./form-element";
import { UimlPartsRepository } from "./uiml-parts";

export class UimlLayoutSerializer {
    static createRoot(): FormElement {
        var formElement = new FormElement(null);
        return formElement;
    }
    static createElement(element: any, parent?: IFormElement): IFormElement {
        var part = UimlPartsRepository.create(element.partclass, element);
        var formElement = new FormElement(parent);
        formElement.content = part;
        UimlLayoutSerializer.createElements(formElement.elements, element.parts, formElement);
        return formElement;
    }
    static createElements(collection: KnockoutObservableArray<IFormElement> | Array<IFormElement>, parts: any[] = [], parent?: IFormElement) {
        parts.forEach(element => {
            collection.push(UimlLayoutSerializer.createElement(element));
        });
    }
    static serialize(root: FormElement): any {
        
    }
}