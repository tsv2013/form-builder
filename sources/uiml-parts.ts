import "./uiml-parts.scss";

export interface IRenderable {
    render(htmlElement: HTMLElement);
    isContainer: boolean;
    hasInnerLayout: boolean;
}

export class UimlPart implements IRenderable {
    public static containerParts = ["layout", "layoutRow", "layoutColumn"];
    public static layoutHolderParts = ["panel"];
    public static layoutConvertableParts = {
        "view": "layout",
        "form": "layout",
        "koWith": "layoutRow",
        "container": "layoutRow",
        "accordion": "layoutRow", // TODO: implement accordion/layout component
        "div": "container",
        "formGroup": "layoutItem"
    };
    public static render: (part: any, container?: HTMLElement) => HTMLElement = undefined;
    private static counter = 1;
    private id: number;
    private _part: any;
    private _partToUse: any;
    constructor(private _partclass: string, ...params: any) {
        this.id = UimlPart.counter++;
        this._part = Object.assign({}, params[0]);
        this._partToUse = Object.assign({}, this._part);
        this._partToUse.partclass = this.partclass;
        this.cssclass = this._part.cssClasses;
    }
    render(htmlElement: HTMLElement) {
        if(UimlPart.render) {
            UimlPart.render(this._partToUse, htmlElement);
        } else {
            htmlElement.className += (" " + this.cssclass);
            if(!this.isContainer) {
                htmlElement.innerHTML = this.partclass + this.id;
            }
        }
    }
    cssclass = "";
    get hasInnerLayout() {
        return UimlPart.layoutHolderParts.indexOf(this.partclass) !== -1;
    }
    get isContainer() {
        return UimlPart.containerParts.indexOf(this.partclass) !== -1;
    }
    get partclass() {
        return UimlPart.layoutConvertableParts[this._partclass] || this._partclass;
    }
    get part() {
        return this._part;
    }
    get data() {
        return this._part.data;
    }
    setParts(parts: any[]) {
        this._part.parts = parts;
        this._partToUse.parts = parts;
    }
}

export class UimlPartsRepository {
    private static parts: {[key: string]: new (...params: any) => UimlPart } = {};
    private static _instance = new UimlPartsRepository();
    static get instance() { return UimlPartsRepository._instance; }
    static register(partclass: string, partConstructor: new () => UimlPart) {
        UimlPartsRepository.parts[partclass] = partConstructor;
    }
    static create(partclass: string, ...params: any) {
        if(!!UimlPartsRepository.parts[partclass]) {
            return new UimlPartsRepository.parts[partclass](...params);
        }
        return new UimlPart(partclass, ...params);
    }
}
