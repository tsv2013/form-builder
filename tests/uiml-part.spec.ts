import { UimlLayoutSerializer } from "../sources/uiml-layout-serializer";
import { UimlPart } from "../sources/uiml-parts";
import { render } from "../uiml";

test("render legacy layout - panel", () => {
    const json = {
        "id": "customer.public.customer.customer_id.payment.public.payment.customer_idPanel",
        "isRelation": true,
        "partclass": "panel",
        "visibleIndex": "customer.public.customer.customer_id.payment.public.payment.customer_id",
        "expanded": false,
        "cssClasses": "relation",
        "text": "payment",
        "customHeaderTemplate": "child-view-context-actions-template",
        "customBodyHeaderTemplate": "child-view-body-actions-template",
        "customHeaderData": "customer.public.customer.customer_id.payment.public.payment.customer_id",
        "parts": [{
            "id": "row2",
            "partclass": "layoutRow",
            "parts": [{
                "id": "customer.public.customer.customer_id.payment.public.payment.customer_idListView",
                "partclass": "child",
                "model": "customer.public.customer.customer_id.payment.public.payment.customer_id"
            }]
        }]
    };
    var elements = [];
    UimlLayoutSerializer.createElements(elements, [json], null);
    var element = elements[0];
    expect(element.content.partclass).toBe("panel");
    expect(element.elements().length).toBe(1);
    expect(element.elements()[0].content.partclass).toBe("layout");
    expect(element.elements()[0].elements().length).toBe(1);
    expect(element.elements()[0].elements()[0].content.partclass).toBe("layoutRow");
    expect(element.elements()[0].elements()[0].elements().length).toBe(1);
    var leafElement = element.elements()[0].elements()[0].elements()[0];
    expect(leafElement.content.partclass).toBe("child");
    expect(leafElement.isContainer).toBeFalsy();

    var container = document.createElement("div");
    UimlPart.render = render;
    leafElement.render(container);
    UimlPart.render = undefined;
    expect(container).toMatchSnapshot();
});