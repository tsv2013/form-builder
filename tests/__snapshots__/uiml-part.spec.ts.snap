// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render legacy layout - panel 1`] = `
<div>
  <component-stub
    class=" "
    params="{ message: 'Error: Component child is not installed.'}"
  />
</div>
`;
